augur/filter.py

```diff
@@ -116,18 +117,59 @@ def register_arguments(parser):
     parser.add_argument('--exclude-ambiguous-dates-by', choices=['any', 'day', 'month', 'year'],
                                 help='Exclude ambiguous dates by day (e.g., 2020-09-XX), month (e.g., 2020-XX-XX), year (e.g., 200X-10-01), or any date fields. An ambiguous year makes the corresponding month and day ambiguous, too, even if those fields have unambiguous values (e.g., "201X-10-01"). Similarly, an ambiguous month makes the corresponding day ambiguous (e.g., "2010-XX-01").')
     parser.add_argument('--query', help="Filter samples by attribute. Uses Pandas Dataframe querying, see https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#indexing-query for syntax.")
-    parser.add_argument('--output', '-o', help="output file", required=True)
+    parser.add_argument('--output', '--output-sequences', '-o', help="filtered sequences in FASTA format")
+    parser.add_argument('--output-metadata', help="metadata for strains that passed filters")
+    parser.add_argument('--output-strains', help="list of strains that passed filters (no header)")
 
     parser.set_defaults(probabilistic_sampling=True)
 
 def run(args):
     '''
     filter and subsample a set of sequences into an analysis set
     '''
+    # Validate arguments before attempting any I/O.
+    # Don't allow sequence output when no sequence input is provided.
+    if args.output and not args.sequences:
+        print(
+            "ERROR: You need to provide sequences to output sequences.",
+            file=sys.stderr)
+        return 1
+
+    # Confirm that at least one output was requested.
+    if not any((args.output, args.output_metadata, args.output_strains)):
+        print(
+            "ERROR: You need to select at least one output.",
+            file=sys.stderr)
+        return 1
+
+    # Don't allow filtering on sequence-based information, if no sequences or
+    # sequence index is provided.
+    SEQUENCE_ONLY_FILTERS = [
+        args.min_length,
+        args.non_nucleotide
+    ]
+    if not args.sequences and not args.sequence_index and any(SEQUENCE_ONLY_FILTERS):
+        print(
+            "ERROR: You need to provide a sequence index or sequences to filter on sequence-specific information.",
+            file=sys.stderr)
+        return 1
+
+    # Load inputs, starting with metadata.
+    try:
+        # Metadata are the source of truth for which sequences we want to keep
```

Comment:
You are right that the way I've written this leads to unexpected behavior because it wasn't intersecting metadata with sequences.

c49a0ec fixes this by explicitly intersecting VCF/FASTA/index strain names with metadata strain names, logging the number of strains filtered by this intersection, and reporting the number in the final filter report. This commit does not replace `all_seqs` with those present in the sequence data, because 1) when sequences are no longer required metadata needs to be the source of truth and 2) we want to track how many sequences are dropped due to missing metadata/sequences. I also just realized that commit only reports the strains in metadata that don't have sequences, but it doesn't report the number of sequences that don't have metadata. I have added this in 902eb94.

In practice, I don't think the resulting behavior changes too much. We always required sequences to have a metadata entry unless they were in the "include" list, so we implicitly filtered sequences on metadata (and wrote out the names of each sequence to standard out).

One side effect of this new implementation is that strains without metadata don't get printed to stdout one by one anymore. I'm convinced that a more useful way to report that information is with an actual structured report output like the one described in #424.